flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java

```diff
@@ -178,9 +178,16 @@ void announceCombinedWatermark() {
                                     aggregator.getAggregatedWatermark().getTimestamp());
                         });
 
-        long maxAllowedWatermark =
-                globalCombinedWatermark.getTimestamp()
-                        + watermarkAlignmentParams.getMaxAllowedWatermarkDrift();
+        long maxAllowedWatermark;
+        try {
+            maxAllowedWatermark =
```

Comment:
nit: maybe we can check with `Long.MAX_VALUE-globalCombinedWatermark.getTimestamp() >= watermarkAlignmentParams.getMaxAllowedWatermarkDrift()`, instead of using the try-catch? 
Or maybe it's better to add a comment explaining why an ArithmeticException can be thrown here.

Reply:
Thanks for the suggestion, but the `Long.MAX_VALUE-globalCombinedWatermark.getTimestamp()` will overflow when `globalCombinedWatermark.getTimestamp()` is negated. If It's necessary to replace the try-catch, we can use: 
```java
        long maxAllowedWatermark =
                globalCombinedWatermark.getTimestamp()
                        + watermarkAlignmentParams.getMaxAllowedWatermarkDrift();
        // check maxAllowedWatermark arithmetic overflow when the source is idle
        if (((globalCombinedWatermark.getTimestamp() ^ maxAllowedWatermark)
                        & (watermarkAlignmentParams.getMaxAllowedWatermarkDrift()
                                ^ maxAllowedWatermark))
                < 0) {
            maxAllowedWatermark = Watermark.MAX_WATERMARK.getTimestamp();
        }
```
WDYT?

Reply:
You are right, and I suppose `globalCombinedWatermark.getTimestamp() > 0 && Long.MAX_VALUE - globalCombinedWatermark.getTimestamp() < watermarkAlignmentParams.getMaxAllowedWatermarkDrift()` would be fine?
I think it better to be clear why and how it is checked. So in fact I would prefer simply adding comments in the try-catch to the bit operation if the simple expression is not enough.

Reply:
I add a comment to explain the ArithmeticException

Reply:
Thanks for the update! LGTM now.